tests/MongoDB.Bson.Tests/HexTests.cs

```diff
@@ -0,0 +1,121 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using FluentAssertions;
+using MongoDB.Shared;
+using Xunit;
+
+namespace MongoDB.Bson.Tests
+{
+    public class HexTests
+    {
+        [Theory]
+        [InlineData(0)]
+        [InlineData('/')]
+        [InlineData(':')]
+        [InlineData('@')]
+        [InlineData('G')]
+        [InlineData('`')]
+        [InlineData('g')]
+        [InlineData(127)]
+        [InlineData(255)]
+        [InlineData(0xffff)]
+        public void IsValidHexDigit_should_return_false_when_value_is_invalid(char c)
+        {
+            var result = Hex.IsValidHexDigit(c);
+
+            result.Should().BeFalse();
+        }
+
+        [Theory]
+        [InlineData('0')]
+        [InlineData('1')]
+        [InlineData('2')]
+        [InlineData('3')]
+        [InlineData('4')]
+        [InlineData('5')]
+        [InlineData('6')]
+        [InlineData('7')]
+        [InlineData('8')]
+        [InlineData('9')]
+        [InlineData('a')]
+        [InlineData('b')]
+        [InlineData('c')]
+        [InlineData('d')]
+        [InlineData('e')]
+        [InlineData('f')]
+        [InlineData('A')]
+        [InlineData('B')]
+        [InlineData('C')]
+        [InlineData('D')]
+        [InlineData('E')]
+        [InlineData('F')]
+        public void IsValidHexDigit_should_return_true_when_value_is_valid(char c)
+        {
+            var result = Hex.IsValidHexDigit(c);
+
+            result.Should().BeTrue();
+        }
+
+        [Theory]
+        [InlineData("x")]
+        [InlineData("0x")]
+        [InlineData("x0")]
+        [InlineData("x00")]
+        [InlineData("0x0")]
+        [InlineData("00x")]
+        public void IsValidHexString_should_return_false_when_value_is_invalid(string value)
+        {
+            var result = Hex.IsValidHexString(value);
+
+            result.Should().BeFalse();
+        }
+
+        [Theory]
+        [InlineData("")]
```

Comment:
Is "" (empty string) is valid hex string?

Reply:
Yes. It would be the hex representation of a zero length byte array.

Reply:
Note: the empty string is NOT a valid `$binary` `$subType` value, but that is tested somewhere else (`HexUtils` knows nothing about `$binary` `$subType`).

Reply:
I agree with 'note'. I asked about the test name which has `IsValidHexString` part. Maybe we may rename it somehow? But I don't see how it can be done, so we can ignore this point.

Reply:
IsValidHexString seems to be the right name to me. It returns true if the string is a valid hex string.

Note: see also the IsValidBinaryDataSubTypeString method in JsonReader.cs. A subtype string is valid if it is a valid hex string of length 1 or 2.